hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -887,94 +851,92 @@ public ClientConnection getConnectionForSql() {
+        if (clusterIdChanged) {
+            logger.warning("Switching from current cluster: " + this.clusterId + " to new cluster: " + newClusterId);
+            client.onClusterRestart();
+        }
+        clusterId = newClusterId;
 
-            TcpClientConnection existingConnection = activeConnections.get(response.memberUuid);
-            if (existingConnection != null) {
-                connection.close("Duplicate connection to same member with uuid : " + response.memberUuid, null);
-                return existingConnection;
-            }
+        connectionAuthenticated(connection, response);
 
-            UUID newClusterId = response.clusterId;
-            if (logger.isFineEnabled()) {
-                logger.fine("Checking the cluster: " + newClusterId + ", current cluster: " + this.clusterId);
-            }
-            // `clusterId` is `null` only at the start of the client.
-            // It is only set in this method below under `clientStateMutex`.
-            // `clusterId` is set by master when a cluster is started.
-            // `clusterId` is not preserved during HotRestart.
-            // In split brain, both sides have the same `clusterId`
-            boolean clusterIdChanged = this.clusterId != null && !newClusterId.equals(this.clusterId);
-            if (clusterIdChanged) {
-                checkClientStateOnClusterIdChange(connection, switchingToNextCluster);
-                logger.warning("Switching from current cluster: " + this.clusterId + " to new cluster: " + newClusterId);
-                client.onClusterRestart();
-            }
-            checkClientState(connection, switchingToNextCluster);
-
-            boolean connectionsEmpty = activeConnections.isEmpty();
-            activeConnections.put(response.memberUuid, connection);
-            if (connectionsEmpty) {
-                // The first connection that opens a connection to the new cluster should set `clusterId`.
-                // This one will initiate `initializeClientOnCluster` if necessary.
-                clusterId = newClusterId;
-                if (clusterIdChanged) {
-                    clientState = ClientState.CONNECTED_TO_CLUSTER;
-                    executor.execute(() -> initializeClientOnCluster(newClusterId));
-                } else {
-                    clientState = ClientState.INITIALIZED_ON_CLUSTER;
-                    fireLifecycleEvent(LifecycleState.CLIENT_CONNECTED);
+        if (clusterIdChanged) {
+            clientState = ClientState.CONNECTED_TO_CLUSTER;
+            try {
+                client.sendStateToCluster();
+                if (logger.isFineEnabled()) {
+                    logger.fine("Client state is sent to cluster: " + clusterId);
                 }
+            } catch (RuntimeException e) {
+                logger.warning("Failure during sending state to the cluster. " + connection, e);
+                activeConnections.remove(connection.getRemoteUuid());
```

Comment:
```suggestion
                activeConnections.remove(connection.getRemoteUuid(), connection);
```

to be sure. and maybe we should check the result of this operation?

Reply:
Fixed it in https://github.com/hazelcast/hazelcast/pull/20806/commits/6161f42da7b5878b0282c99067f500ff3bb9654b